File: Sources/TamaCore/Usage/CostEstimator.swift
import Foundation

public struct Rates: Sendable, Equatable {
    public let input: Double       // USD per 1M tokens
    public let output: Double
    public let cacheRead: Double
    public let cacheWrite: Double    // 5-minute cache write (1.25× input)
    public let cacheWrite1h: Double  // 1-hour cache write (2× input); defaults to cacheWrite if omitted
    public init(input: Double, output: Double, cacheRead: Double, cacheWrite: Double, cacheWrite1h: Double? = nil) {
        self.input = input; self.output = output; self.cacheRead = cacheRead
        self.cacheWrite = cacheWrite; self.cacheWrite1h = cacheWrite1h ?? cacheWrite
    }
}

extension Rates: Decodable {
    private enum CodingKeys: String, CodingKey { case input, output, cacheRead, cacheWrite, cacheWrite1h }
    public init(from decoder: Decoder) throws {
        let c = try decoder.container(keyedBy: CodingKeys.self)
        // cacheWrite1h is optional in prices.json; absent → priced like a 5-minute write (no regression).
        self.init(input: try c.decode(Double.self, forKey: .input),
                  output: try c.decode(Double.self, forKey: .output),
                  cacheRead: try c.decode(Double.self, forKey: .cacheRead),
                  cacheWrite: try c.decode(Double.self, forKey: .cacheWrite),
                  cacheWrite1h: try c.decodeIfPresent(Double.self, forKey: .cacheWrite1h))
    }
}

/// One provider's rates: a `default` (used when the model is unknown) plus optional per-model-tier
/// overrides keyed by a substring of the model id ("opus"/"sonnet"/"haiku"/"fable"). Claude prices
/// each tier very differently (Opus ≈ 5× Haiku), so pricing the whole provider at one flat rate is
/// wrong — `default` is only the fallback for ids that match no tier.
public struct ProviderRates: Sendable, Equatable {
    public let `default`: Rates
    public let models: [String: Rates]
    public init(default d: Rates, models: [String: Rates] = [:]) {
        self.default = d; self.models = models
    }
}

extension ProviderRates: Decodable {
    private enum CodingKeys: String, CodingKey { case `default`, models }
    public init(from decoder: Decoder) throws {
        // Preferred shape: { "default": {...}, "models": { "opus": {...}, ... } }.
        if let c = try? decoder.container(keyedBy: CodingKeys.self), c.contains(.default) {
            let d = try c.decode(Rates.self, forKey: .default)
            let m = (try? c.decode([String: Rates].self, forKey: .models)) ?? [:]
            self.init(default: d, models: m)
        } else {
            // Back-compat: a bare flat `Rates` object becomes the default with no per-model overrides.
            self.init(default: try Rates(from: decoder), models: [:])
        }
    }
}

public struct CostEstimator: Sendable {
    private let rates: [Provider: ProviderRates]

    /// Model-aware construction (default + per-tier overrides).
    public init(providerRates: [Provider: ProviderRates]) { self.rates = providerRates }

    /// Flat, model-agnostic construction — every model for a provider is priced the same.
    public init(rates: [Provider: Rates]) { self.rates = rates.mapValues { ProviderRates(default: $0) } }

    /// Loads rates from bundled prices.json, falling back to built-in (model-aware) defaults.
    public init() { self.rates = CostEstimator.loadBundledRates() }

    /// Cost of a breakdown for a provider, model-agnostic (uses the provider's default rate).
    public func cost(_ b: TokenBreakdown, provider: Provider) -> Double {
        cost(b, provider: provider, model: nil)
    }

    /// Cost of a breakdown for a provider at the rate for `model` (its tier), or the default rate
    /// when `model` is nil or matches no known tier.
    public func cost(_ b: TokenBreakdown, provider: Provider, model: String?) -> Double {
        guard let pr = rates[provider] else { return 0 }
        let r = Self.rate(for: model, in: pr)
        let write1h = min(b.cacheWrite1h, b.cacheWrite)      // 1-hour portion of cache writes
        let write5m = max(0, b.cacheWrite - write1h)         // remainder is 5-minute
        return Double(b.input)     / 1_000_000 * r.input
             + Double(b.output)    / 1_000_000 * r.output
             + Double(b.cacheRead) / 1_000_000 * r.cacheRead
             + Double(write5m)     / 1_000_000 * r.cacheWrite
             + Double(write1h)     / 1_000_000 * r.cacheWrite1h
    }

    /// Pick the tier rate whose key appears in the (lowercased) model id; else the default.
    static func rate(for model: String?, in pr: ProviderRates) -> Rates {
        guard !pr.models.isEmpty, let m = model?.lowercased() else { return pr.default }
        for (key, r) in pr.models where m.contains(key) { return r }
        return pr.default
    }

    /// Flat per-provider fallback rates (the default tier each provider falls back to). Claude's
    /// default is the Opus tier — the flagship and what Tama's own sessions run on.
    public static let defaultRates: [Provider: Rates] = [
        .claudeCode: Rates(input: 5.0, output: 25.0, cacheRead: 0.5, cacheWrite: 6.25, cacheWrite1h: 10.0),
        .codex:      Rates(input: 2.5, output: 10.0, cacheRead: 0.25, cacheWrite: 2.5),
    ]

    /// Per-tier Claude rates (USD per 1M tokens). cacheRead ≈ 0.1× input, cacheWrite (5-min) ≈ 1.25×
    /// input, cacheWrite1h (1-hour) ≈ 2× input. Claude Code writes to the 1-hour cache.
    static let claudeModelRates: [String: Rates] = [
        "opus":   Rates(input: 5.0,  output: 25.0, cacheRead: 0.5,  cacheWrite: 6.25,  cacheWrite1h: 10.0),
        "sonnet": Rates(input: 3.0,  output: 15.0, cacheRead: 0.3,  cacheWrite: 3.75,  cacheWrite1h: 6.0),
        "haiku":  Rates(input: 1.0,  output: 5.0,  cacheRead: 0.1,  cacheWrite: 1.25,  cacheWrite1h: 2.0),
        "fable":  Rates(input: 10.0, output: 50.0, cacheRead: 1.0,  cacheWrite: 12.5,  cacheWrite1h: 20.0),
    ]

    /// Built-in model-aware fallback used when prices.json is absent or unparseable.
    static var defaultProviderRates: [Provider: ProviderRates] {
        [
            .claudeCode: ProviderRates(default: defaultRates[.claudeCode]!, models: claudeModelRates),
            .codex:      ProviderRates(default: defaultRates[.codex]!),
        ]
    }

    /// Searches likely bundle locations WITHOUT touching Bundle.module (which calls fatalError
    /// internally when the resource bundle is missing — uncatchable by try?).
    private static func bundledPricesURL() -> URL? {
        let bundleName = "Tama_TamaCore.bundle"
        let bases: [URL] = [
            Bundle.main.resourceURL,
            Bundle.main.bundleURL,
            Bundle.main.bundleURL.appendingPathComponent("Contents/Resources"),
        ].compactMap { $0 }
        for base in bases {
            let inBundle = base.appendingPathComponent(bundleName).appendingPathComponent("prices.json")
            if FileManager.default.fileExists(atPath: inBundle.path) { return inBundle }
            let direct = base.appendingPathComponent("prices.json")
            if FileManager.default.fileExists(atPath: direct.path) { return direct }
        }
        return nil
    }

    static func loadBundledRates() -> [Provider: ProviderRates] {
        guard let url = bundledPricesURL(),
              // Read the bundled prices.json through SafeFileReader (size cap + symlink/regular-file
              // check), keeping every disk read on the one hardened path. 1 MiB is ample headroom.
              let data = SafeFileReader.data(at: url, maxBytes: 1 << 20),
              let raw = try? JSONDecoder().decode([String: ProviderRates].self, from: data) else {
            return defaultProviderRates
        }
        var out: [Provider: ProviderRates] = [:]
        for (k, v) in raw { if let p = Provider(rawValue: k) { out[p] = v } }
        return out.isEmpty ? defaultProviderRates : out
    }
}

File: windows/Tama.Core/CostEstimator.cs
using System.Reflection;
using System.Text.Json;

namespace Tama.Core;

/// <summary>USD per 1M tokens. CacheWrite = 5-minute TTL rate; CacheWrite1h = 1-hour TTL rate
/// (defaults to CacheWrite when a price file omits it).</summary>
public readonly record struct Rates(
    double Input, double Output, double CacheRead, double CacheWrite, double CacheWrite1h)
{
    public Rates(double Input, double Output, double CacheRead, double CacheWrite)
        : this(Input, Output, CacheRead, CacheWrite, CacheWrite) { }
}

/// <summary>One provider's rates: a Default (unknown model) plus per-tier overrides keyed by a
/// substring of the model id ("opus"/"sonnet"/…). Mirror of the Swift ProviderRates.</summary>
public sealed record ProviderRates(Rates Default, IReadOnlyDictionary<string, Rates> Models);

/// <summary>Estimates pay-as-you-go API cost of token usage. Mirror of the Swift CostEstimator.</summary>
public sealed class CostEstimator
{
    private readonly IReadOnlyDictionary<Provider, ProviderRates> _rates;

    public CostEstimator(IReadOnlyDictionary<Provider, ProviderRates> providerRates) =>
        _rates = providerRates;

    /// <summary>Loads rates from the embedded prices.json, falling back to built-in defaults.</summary>
    public CostEstimator() => _rates = LoadEmbeddedRates();

    /// <summary>Cost of a breakdown at the rate for the model's tier, or the provider default.</summary>
    public double Cost(TokenBreakdown b, Provider provider, string? model = null)
    {
        if (!_rates.TryGetValue(provider, out var pr)) return 0;
        var r = RateFor(model, pr);
        var write1h = Math.Min(b.CacheWrite1h, b.CacheWrite);   // 1-hour portion of cache writes
        var write5m = Math.Max(0, b.CacheWrite - write1h);      // remainder is 5-minute
        return b.Input / 1e6 * r.Input
             + b.Output / 1e6 * r.Output
             + b.CacheRead / 1e6 * r.CacheRead
             + write5m / 1e6 * r.CacheWrite
             + write1h / 1e6 * r.CacheWrite1h;
    }

    /// <summary>Pick the tier whose key appears in the lowercased model id; else the default.</summary>
    public static Rates RateFor(string? model, ProviderRates pr)
    {
        if (pr.Models.Count == 0 || model is null) return pr.Default;
        var m = model.ToLowerInvariant();
        foreach (var (key, r) in pr.Models)
            if (m.Contains(key, StringComparison.Ordinal)) return r;
        return pr.Default;
    }

    // Built-in model-aware fallback used when prices.json is absent or unparseable.
    // Claude's default is the Opus tier; per-tier table mirrors the Swift claudeModelRates.
    internal static IReadOnlyDictionary<Provider, ProviderRates> DefaultProviderRates => new Dictionary<Provider, ProviderRates>
    {
        [Provider.ClaudeCode] = new(new Rates(5.0, 25.0, 0.5, 6.25, 10.0), new Dictionary<string, Rates>
        {
            ["opus"] = new(5.0, 25.0, 0.5, 6.25, 10.0),
            ["sonnet"] = new(3.0, 15.0, 0.3, 3.75, 6.0),
            ["haiku"] = new(1.0, 5.0, 0.1, 1.25, 2.0),
            ["fable"] = new(10.0, 50.0, 1.0, 12.5, 20.0),
        }),
        [Provider.Codex] = new(new Rates(2.5, 10.0, 0.25, 2.5), new Dictionary<string, Rates>()),
    };

    private static IReadOnlyDictionary<Provider, ProviderRates> LoadEmbeddedRates()
    {
        try
        {
            using var stream = Assembly.GetExecutingAssembly()
                .GetManifestResourceStream("Tama.Core.prices.json");
            if (stream is null) return DefaultProviderRates;
            using var doc = JsonDocument.Parse(stream);
            if (doc.RootElement.ValueKind != JsonValueKind.Object) return DefaultProviderRates;
            var result = new Dictionary<Provider, ProviderRates>();
            foreach (var prop in doc.RootElement.EnumerateObject())
            {
                var provider = prop.Name switch
                {
                    "claudeCode" => Provider.ClaudeCode,
                    "codex" => Provider.Codex,
                    "gemini" => Provider.Gemini,
                    "antigravity" => Provider.Antigravity,
                    _ => (Provider?)null,
                };
                if (provider is null || ParseProviderRates(prop.Value) is not { } pr) continue;
                result[provider.Value] = pr;
            }
            return result.Count > 0 ? result : DefaultProviderRates;
        }
        catch (JsonException) { return DefaultProviderRates; }
    }

    private static ProviderRates? ParseProviderRates(JsonElement el)
    {
        if (el.ValueKind != JsonValueKind.Object) return null;
        // Preferred shape: { "default": {...}, "models": {...} }; back-compat: a bare Rates object.
        if (el.TryGetProperty("default", out var d))
        {
            if (ParseRates(d) is not { } def) return null;
            var models = new Dictionary<string, Rates>();
            if (el.TryGetProperty("models", out var ms) && ms.ValueKind == JsonValueKind.Object)
                foreach (var m in ms.EnumerateObject())
                    if (ParseRates(m.Value) is { } r) models[m.Name] = r;
            return new ProviderRates(def, models);
        }
        return ParseRates(el) is { } bare
            ? new ProviderRates(bare, new Dictionary<string, Rates>()) : null;
    }

    private static Rates? ParseRates(JsonElement el)
    {
        if (el.ValueKind != JsonValueKind.Object) return null;
        double? Get(string name) =>
            el.TryGetProperty(name, out var v) && v.ValueKind == JsonValueKind.Number
                ? v.GetDouble() : null;
        if (Get("input") is not { } i || Get("output") is not { } o
            || Get("cacheRead") is not { } cr || Get("cacheWrite") is not { } cw) return null;
        // cacheWrite1h optional: absent → priced like a 5-minute write (no regression).
        return Get("cacheWrite1h") is { } cw1h
            ? new Rates(i, o, cr, cw, cw1h) : new Rates(i, o, cr, cw);
    }
}
